File: leetcode/Python/0833-m-find-and-replace-in-string.py
class Solution_1:
    def findReplaceString(self, S: str, indexes: List[int], sources: List[str], targets: List[str]) -> str:
        res = ""
        mapper = {}
        for i in range(len(indexes)):
            mapper[indexes[i]] = (sources[i], targets[i])

        i = 0
        while i < len(S):
            if i in mapper and S[i: i + len(mapper[i][0])] == mapper[i][0]:
                res += mapper[i][1]
                i += len(mapper[i][0])
            else:
                res += S[i]
                i += 1
        return res

class Solution_2:
    def findReplaceString(self, S: str, indexes: List[int], sources: List[str], targets: List[str]) -> str:
        for i, s, t in sorted(zip(indexes, sources, targets), reverse=True):
            if S[i:i + len(s)] == s:
                S = S[:i] + t + S[i + len(s):]
            else:
                S = S
        return S

File: leetcode/C++/0833-m-find-and-replace-in-string.cpp
class Solution {
public:
    string findReplaceString(string S, vector<int>& indexes, vector<string>& sources, vector<string>& targets) {
        map<int, pair<string, string>> mp;
        for (int i = 0; i < indexes.size(); i++)
            mp[indexes[i]] = {sources[i], targets[i]};
        
        int p = 0;
        string res;
        while (p < S.length()) {
            if (mp.find(p) != mp.end() && S.substr(p, mp[p].first.size()) == mp[p].first) {
                res += mp[p].second;
                p += mp[p].first.size();
            } else {
                res += S[p];
                p++;
            }
        }
        return res;
    }
};